sample/sample.go

```diff
@@ -43,65 +54,162 @@ func (s *SamplerFactory) updatePeerCounts() {
+		dynsamplerInstance := getSharedDynsampler(s, dynsamplerKey, c, createDynForTotalThroughputSampler)
+		// Store goal throughput config under mutex protection
+		s.mutex.Lock()
+		s.goalThroughputConfigs[dynsamplerKey] = c.GoalThroughputPerSec
+		s.mutex.Unlock()
+		sampler = &TotalThroughputSampler{Config: c, Logger: s.Logger, Metrics: s.Metrics, dynsampler: dynsamplerInstance}
 	case *config.EMAThroughputSamplerConfig:
-		sampler = &EMAThroughputSampler{Config: c, Logger: s.Logger, Metrics: s.Metrics}
+		dynsamplerKey := fmt.Sprintf("%s:emathroughput:%d:%v", keyPrefix, c.GoalThroughputPerSec, c.FieldList)
+		dynsamplerInstance := getSharedDynsampler(s, dynsamplerKey, c, createDynForEMAThroughputSampler)
+		s.mutex.Lock()
+		s.goalThroughputConfigs[dynsamplerKey] = c.GoalThroughputPerSec
+		s.mutex.Unlock()
+		sampler = &EMAThroughputSampler{Config: c, Logger: s.Logger, Metrics: s.Metrics, dynsampler: dynsamplerInstance}
 	case *config.WindowedThroughputSamplerConfig:
-		sampler = &WindowedThroughputSampler{Config: c, Logger: s.Logger, Metrics: s.Metrics}
+		dynsamplerKey := fmt.Sprintf("%s:windowedthroughput:%d:%v", keyPrefix, c.GoalThroughputPerSec, c.FieldList)
+		dynsamplerInstance := getSharedDynsampler(s, dynsamplerKey, c, createDynForWindowedThroughputSampler)
+		s.mutex.Lock()
+		s.goalThroughputConfigs[dynsamplerKey] = c.GoalThroughputPerSec
+		s.mutex.Unlock()
+		sampler = &WindowedThroughputSampler{Config: c, Logger: s.Logger, Metrics: s.Metrics, dynsampler: dynsamplerInstance}
 	default:
 		s.Logger.Error().Logf("unknown sampler type %T. Exiting.", c)
 		os.Exit(1)
+		return nil
 	}
 
 	err := sampler.Start()
 	if err != nil {
-		s.Logger.Debug().WithField("dataset", samplerKey).Logf("failed to start sampler")
+		s.Logger.Debug().WithField("dataset", keyPrefix).Logf("failed to start sampler")
 		return nil
 	}
 
-	s.Logger.Debug().WithField("dataset", samplerKey).Logf("created implementation for sampler type %T", c)
-	// call this every time we add a sampler
-	s.samplers = append(s.samplers, sampler)
+	s.Logger.Debug().WithField("dataset", keyPrefix).Logf("created implementation for sampler type %T", c)
+	// Update peer counts after creating a sampler
 	s.updatePeerCounts()
 
 	return sampler
 }
 
+// GetSamplerImplementationForKey returns the sampler implementation for the given
+// samplerKey (dataset for legacy keys, environment otherwise), or nil if it is not defined
+func (s *SamplerFactory) GetSamplerImplementationForKey(samplerKey string) Sampler {
+	c, _ := s.Config.GetSamplerConfigForDestName(samplerKey)
+
+	return s.createSampler(c, "")
+}
+
+// GetDownstreamSampler creates a downstream sampler for use in rules-based sampling,
+// ensuring it participates in the shared dynsampler system
+func (s *SamplerFactory) GetDownstreamSampler(
+	parentSamplerKey string,
+	downstreamConfig *config.RulesBasedDownstreamSampler,
+) Sampler {
+	keyPrefix := fmt.Sprintf("rules:%s:", parentSamplerKey)
```

Comment:
I wonder if it would be more clear if we always put environment or dataset name first in the dynsampler keys. Instead of `rules:production:totoalthroughput` and `production:totalthroughput`, we can have `production:rules:totalthroughput`

Reply:
I see where you're coming from but doing that complicates creating these keys, for what it's ultimately an aesthetic choice. I don't think it's worthwhile.